### PR TITLE
[WIP] Add support for `approx.auto_quantize`

### LINKED ISSUE #3: Implement `approx.auto_cast_all()` and test
The ability to run `approx.auto_cast_all()` to auto-cast everything in the model (similar to what PyTorch has)

### LINKED ISSUE #16: Implement `approx.auto_quantize()` function
**Describe the solution you'd like**

Adds support for a public API function, `approx.auto_quantize(model, pretrained: bool)`

This accepts some model and automatically quantizes it

**Additional context**

This replaces #3 temporarily for a nicer user-facing API

Additional improvements could include supporting an `@auto_quantize` annotation over some given model

### PR DESCRIPTION
**Pull request by bushshrub**
This PR adds support for `approx.auto_quantize` as covered in #16.

This function takes some model (currently only torch module) and returns a quantized version of that model

Also changes the function signature of `auto_set_backend()` to modify a library variable instead

Closes: #16 

May also close #3

### CODE REVIEW

File: examples/scripts/mnist_torch.py [hunk 1]
```diff
@@ -32,26 +32,35 @@ def forward(self, x):
...
-test_loader = data.DataLoader(mnist_test, batch_size=512, shuffle=True)
+    train_loader = data.DataLoader(mnist_train, batch_size=512, shuffle=True)
+    test_loader = data.DataLoader(mnist_test, batch_size=512, shuffle=True)
 
-model = BasicNN()
-model.to("cuda")
-optim = torch.optim.SGD(model.parameters(), lr=0.01)
-loss = nn.CrossEntropyLoss().to("cuda")
+    model = BasicNN()
+    model = approx.auto_quantize(model, pretrained=True)
```
Comment (xiurobert) [inline examples/scripts/mnist_torch.py:48]: Nit: extra line spacing

File: examples/scripts/mnist_torch.py [hunk 2]
```diff
@@ -32,26 +32,35 @@ def forward(self, x):
...
+    # model.to("cuda")
+    optim = torch.optim.SGD(model.parameters(), lr=0.001)
+    loss = nn.CrossEntropyLoss().to("cuda")
+
+    model.train()
+    num_epochs = 2
     for epoch_it in range(num_epochs):
         prog = tqdm(enumerate(train_loader), total=len(train_loader))
         for batch_num, (features, labels) in prog:
+            # approx.auto_cast_all(features, labels)
```
Comment (bushshrub) [inline examples/scripts/mnist_torch.py]: Remove unnecessary comment

File: examples/scripts/mnist_torch.py (hunk 1 above)
Comment (bushshrub) [inline examples/scripts/mnist_torch.py]: Extra spacing

File: examples/scripts/mnist_torch.py [hunk 3]
```diff
@@ -32,26 +32,35 @@ def forward(self, x):
...
-model = BasicNN()
-model.to("cuda")
-optim = torch.optim.SGD(model.parameters(), lr=0.01)
-loss = nn.CrossEntropyLoss().to("cuda")
+    model = BasicNN()
+    model = approx.auto_quantize(model, pretrained=True)
 
-with approx.auto_cast_all():
-    num_epochs = 10
+    # qmodel = torch.quantization.quantize_dynamic(model, {nn.Linear, nn.Conv2d}, dtype=torch.qint8)
```
Comment (bushshrub) [inline examples/scripts/mnist_torch.py]: Remove this

File: examples/scripts/mnist_torch.py [hunk 4]
```diff
@@ -32,26 +32,35 @@ def forward(self, x):
...
+    # qmodel.to('cuda')
+    print(model)
+    exit(0)
+
+    # model.to("cuda")
```
Comment (bushshrub) [inline examples/scripts/mnist_torch.py]: Fix this

File: tests/integration/core/test_api.py [hunk 5]
```diff
@@ -5,12 +5,14 @@
 
 def test_auto_set_backend():
     """Tests that the auto_set_backend function works as expected."""
-    from approx.core.api import auto_set_backend
-    from approx.core.backend._backend import BackendType
+    # from approx.core.api import auto_set_backend
```
Comment (bushshrub) [inline tests/integration/core/test_api.py]: Unnecessary

File: approx/core/backend/_backend.py [hunk 6]
```diff
@@ -26,30 +28,60 @@ def __init__(self, backend):
...
+        """
+        if pretrained:
+            qmodel = self.module.quantization.quantize_dynamic(
+                model,
+                {
+                    # todo: this is definitely not enough
+                    self.module.nn.Linear,
+                    self.module.nn.Conv2d,
+                },
+                dtype=self.module.qint8,
```
Comment (bushshrub) [inline approx/core/backend/_backend.py]: More quantization options?

@sudomaze

**Review by bushshrub — COMMENTED** [review]
Few more things

File: approx/core/backend/_backend.py (hunk 6 above)
Comment (ma7dev) [reply approx/core/backend/_backend.py]: I am adding more per [1](https://pytorch.org/docs/stable/_modules/torch/ao/quantization/quantize.html#quantize_dynamic) and [2](https://pytorch.org/docs/stable/quantization.html#quantization-customizations). Here is the current list that was added: 

```python
    self.module.nn.Linear,
    self.module.nn.LSTM,
    self.module.nn.GRU,
    self.module.nn.LSTMCell,
    self.module.nn.RNNCell,
    self.module.nn.GRUCell,
    self.module.nn.Conv2d,
    self.module.nn.Conv3d,
```

I will add some unit tests to ensure the quantization is working as expected. We will need to do a similar mapping in other backends.

Comment (bushshrub) [reply approx/core/backend/_backend.py]: SGTM. I was mainly working on PyTorchBackend until we get that decently well working, before we support others

**Comment by bushshrub** [comment]
Blocking, waiting for #22 to hit develop

**Review by ma7dev — APPROVED** [review]

**Pull request merged**